collector/perf_linux.go

```diff
@@ -35,37 +42,85 @@ func init() {
 // settings not all profiler values may be exposed on the target system at any
 // given time.
 type perfCollector struct {
-	perfHwProfilers    map[int]perf.HardwareProfiler
-	perfSwProfilers    map[int]perf.SoftwareProfiler
-	perfCacheProfilers map[int]perf.CacheProfiler
-	desc               map[string]*prometheus.Desc
+	hwProfilerCpuMap    map[*perf.HardwareProfiler]int
+	swProfilerCpuMap    map[*perf.SoftwareProfiler]int
+	cacheProfilerCpuMap map[*perf.CacheProfiler]int
+	perfHwProfilers     map[int]*perf.HardwareProfiler
+	perfSwProfilers     map[int]*perf.SoftwareProfiler
+	perfCacheProfilers  map[int]*perf.CacheProfiler
+	desc                map[string]*prometheus.Desc
+}
+
+func isValidCPUString(cpus *string) bool {
+	if !strings.Contains(*cpus, "-") || strings.Count(*cpus, "-") != 1 {
+		return false
+	}
+
+	return true
 }
 
 // NewPerfCollector returns a new perf based collector, it creates a profiler
 // per CPU.
 func NewPerfCollector() (Collector, error) {
 	collector := &perfCollector{
-		perfHwProfilers:    map[int]perf.HardwareProfiler{},
-		perfSwProfilers:    map[int]perf.SoftwareProfiler{},
-		perfCacheProfilers: map[int]perf.CacheProfiler{},
+		perfHwProfilers:     map[int]*perf.HardwareProfiler{},
+		perfSwProfilers:     map[int]*perf.SoftwareProfiler{},
+		perfCacheProfilers:  map[int]*perf.CacheProfiler{},
+		hwProfilerCpuMap:    map[*perf.HardwareProfiler]int{},
+		swProfilerCpuMap:    map[*perf.SoftwareProfiler]int{},
+		cacheProfilerCpuMap: map[*perf.CacheProfiler]int{},
+	}
+
+	start := 0
+	ncpus := 0
+	var err error
+
+	if perfCpusFlag == nil || *perfCpusFlag == "" {
+		start = 0
+		ncpus = runtime.NumCPU() - 1
+	} else if !isValidCPUString(perfCpusFlag) {
+		return nil, errors.New("--collector.perf.cpus flag value is invalid, it must be a range (e.g. 2-6)")
+	} else {
+		cpuRange := strings.Split(*perfCpusFlag, "-")
+		start, err = strconv.Atoi(cpuRange[0])
+		if err != nil {
+			start = 0
+		}
+
+		ncpus, err = strconv.Atoi(cpuRange[1])
+		if err != nil {
+			ncpus = runtime.NumCPU() - 1
+		}
 	}
-	ncpus := runtime.NumCPU()
-	for i := 0; i < ncpus; i++ {
+
+	for i, idx := start, 0; i <= ncpus; i, idx = i+1, idx+1 {
 		// Use -1 to profile all processes on the CPU, see:
 		// man perf_event_open
-		collector.perfHwProfilers[i] = perf.NewHardwareProfiler(-1, i)
-		if err := collector.perfHwProfilers[i].Start(); err != nil {
+		p := perf.NewHardwareProfiler(-1, i)
```

Comment:
Let's rename the p's here:
- ph = hardware profiler
- ps = software profiler
- pc = cache profiler